std/xml.d

```diff
@@ -2045,8 +2045,8 @@ class ElementParser
 
                     immutable(char)* p = startTag.tagString.ptr
                         + startTag.tagString.length;
-                    immutable(char)* q = tag_.tagString.ptr;
-                    text = decode(p[0..(q-p)], DecodeMode.LOOSE);
+                    immutable end = &tag_.tagString[0] - (&startTag.tagString[0] + startTag.tagString.length);
+                    text = decode(p[0 .. end], DecodeMode.LOOSE);
```

Comment:
I really do not like this piece of code (both the original version and the revised version).  In the original code, the construct `p[0 .. (q-p)]` implies that p must be smaller than q, but p points to the end of `startTag.tagString`. So q must lie *outside* of `startTag.tagString`.  And then it takes a slice from p to q, which is a slice that lies *outside* of `startTag.tagString`.  The only way this can even remotely be valid is that q points to a later part of the same input as `startTag.tagString` is a substring of.  While this is probably true, this makes this code very fragile and highly-dependent on the behaviour of the surrounding code.  I'm not surprised the compiler is not happy with it.

However, I *am* surprised that the compiler is OK with the revised version of the code, that essentially *does the same thing*.  How it can decide that the revised code is `@safe` is beyond me, unless it's a loophole in the compiler.  Note that there is *no* bounds check on the slice `p[0 .. end]` because `p` is a raw pointer.  We don't even know (locally) that `tag_.tagString` and `startTag.tagString` are even remotely related to each other, so any bounds check introduced by indexing them with 0 does not, AFAICT, guarantee anything at all.  IOW this code is still (locally) unsafe.  Unless the compiler is doing aliasing analysis on the entire function to guarantee that the slices are valid, but I have my doubts, because in that case, the original code should have been accepted too.  So I'm not sure how/why the revised code makes a difference, and I'm uncomfortable about the compiler accepting the new code in spite of the semantics being essentially identical (i.e., taking what amounts to an out-of-bounds slice of `startTag.tagString`).